File: LeetCode/450. Delete Node in a BST.cpp
/***  KNOWLEDGE IS POWER  ***/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if (root) {
            if (root->val > key)
                root->left = deleteNode(root->left, key);
            else if (root->val < key)
                root->right = deleteNode(root->right, key);
            else {
                if (root->left == root->right)
                    return NULL;
                else if (root->left && root->right) {
                    TreeNode* tmp = root->left;
                    while (tmp->right)
                        tmp = tmp->right;
                    root->val = tmp->val;
                    root->left = deleteNode(root->left, root->val);
                }
                else
                    return root->left ? root->left : root->right;
            }
        }
        return root ;
    }
};

File: LeetCode/450. Delete Node in a BST.java
/***  KNOWLEDGE IS POWER  ***/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root != null) {
            if (root.val > key)
                root.left = deleteNode(root.left, key);
            else if (root.val < key)
                root.right = deleteNode(root.right, key);
            else {
                if (root.left == root.right)
                    return null;
                else if (root.left != null && root.right != null) {
                    TreeNode tmp = root.left;
                    while (tmp.right != null)
                        tmp = tmp.right;
                    root.val = tmp.val;
                    root.left = deleteNode(root.left, root.val);
                }
                else
                    return root.left != null ? root.left : root.right;
            }
        }
        return root;
    }
}
